### PR TITLE
 core/mvcc: Implement Hekaton commit dependency tracking 

### PR DESCRIPTION
**Pull request by pereman2**
## Description
Fixes https://github.com/tursodatabase/turso/issues/5136:

Implement the register-and-report commit dependency protocol from the
Hekaton paper (Section 2.7) to fix partial commit visibility.

During commit, row timestamps are converted from TxID to Timestamp one
by one while the transaction is in Preparing state. Without dependency
tracking, concurrent readers could see partial commits — some rows
visible (already converted to Timestamp) and others not (still TxID
pointing to a Preparing tx), violating snapshot isolation.

The fix changes is_begin_visible to speculatively read from Preparing
transactions when begin_ts >= end_ts, and registers a commit dependency
via the Hekaton protocol:

- CommitDepCounter (AtomicU64): counts unresolved dependencies; must
  reach zero before a transaction can commit.
- AbortNow (AtomicBool): set by a depended-on transaction on abort to
  cascade the abort to dependents.
- CommitDepSet (Mutex<Vec<TxID>>): stores IDs of transactions that
  depend on this one; drained during commit/abort postprocessing.

New WaitForDependencies commit state sits between validation and
timestamp postprocessing, yielding until all dependencies resolve.

---

> [!NOTE]
> **High Risk**
> Touches core MVCC visibility/commit state machine and adds cross-transaction dependency tracking with new waiting/abort behavior, which is concurrency-sensitive and could introduce subtle correctness or liveness regressions.
> 
> **Overview**
> Fixes partial-commit visibility under snapshot isolation by implementing Hekaton §2.7 commit dependencies: transactions can now *speculatively* read/ignore versions from `Preparing` transactions and register a dependency via `commit_dep_counter`, `abort_now`, and `commit_dep_set`.
> 
> The commit state machine adds `WaitForDependencies` between validation and timestamp postprocessing, yielding until dependencies resolve, and cascades aborts/notifications by draining `commit_dep_set` on commit/rollback (with a new `LimboError::CommitDependencyAborted`). Extensive new tests cover speculative read/ignore, cascade abort, read-only edge cases, and a regression reproducer for the prior partial-visibility bug.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 22c0deb2ddcbd04c6480456038782aa51fa6ff15. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: core/mvcc/database/mod.rs [hunk 1]
```diff
@@ -999,7 +1016,7 @@ impl<Clock: LogicalClock> StateTransition for CommitStateMachine<Clock> {
                     """
                 **  If you're wondering when a speculative read happens, here you go:
                 **  Case 1: speculative read of TB:
-                    """
+                    """Expand commentComment on line L983Resolved
```
Comment (avinassh) [inline core/mvcc/database/mod.rs]: unnecessary comment creeped in

File: core/mvcc/database/mod.rs [hunk 2]
```diff
@@ -391,6 +391,16 @@ pub struct Transaction {
     savepoint_stack: RwLock<Vec<Savepoint>>,
     /// True when this transaction currently holds the serialized logical-log commit lock.
     pager_commit_lock_held: AtomicBool,
+    /// Number of unresolved commit dependencies (must reach 0 before commit).
+    /// Hekaton Section 2.7: "A transaction cannot commit until this counter is zero."
+    commit_dep_counter: AtomicU64,
```
Comment (avinassh) [inline core/mvcc/database/mod.rs:458]:
```suggestion
    /// Number of unresolved commit dependencies (must reach 0 before commit).
    /// i.e. the number of transactions this transaction is dependent on and waiting 
    /// for them to commit (or abort).
    /// Hekaton Section 2.7: "A transaction cannot commit until this counter is zero."
    commit_dep_counter: AtomicU64,
```

File: core/mvcc/database/mod.rs [hunk 3]
```diff
@@ -1035,6 +1063,22 @@ impl<Clock: LogicalClock> StateTransition for CommitStateMachine<Clock> {
...
+                        // and go straight to WaitForDependencies.
+                        self.state = CommitState::WaitForDependencies { end_ts };
+                        return Ok(TransitionResult::Continue);
+                    }
+                    // Check abort_now AFTER counter: rollback_tx stores abort_now
+                    // (Release) before fetch_sub (AcqRel). Once counter == 0, all
+                    // decrements have completed and the abort_now flag is visible.
+                    if tx.abort_now.load(Ordering::Acquire) {
+                        return Err(LimboError::CommitDependencyAborted);
+                    }
```
Comment (avinassh) [inline core/mvcc/database/mod.rs:1291]: I suggest flipping the order. We should check for `abort_now` first, so that we can abort eagerly. And then wait for `commit_dep_counter` to be zero

File: core/mvcc/database/mod.rs [hunk 4]
```diff
@@ -1035,6 +1063,22 @@ impl<Clock: LogicalClock> StateTransition for CommitStateMachine<Clock> {
                     b.table_id.cmp(&a.table_id).then(a.row_id.cmp(&b.row_id))
                 });
                 if self.write_set.is_empty() {
```
Comment (avinassh) [inline core/mvcc/database/mod.rs:1267]: a read txn should not have any dependents. let's assert that `commit_dep_set` is zero.

File: core/mvcc/database/mod.rs [hunk 5]
```diff
@@ -1068,7 +1112,58 @@ impl<Clock: LogicalClock> StateTransition for CommitStateMachine<Clock> {
...
+                // Check abort_now AFTER counter reaches 0. Memory ordering:
+                // rollback_tx does abort_now.store(true, Release) BEFORE
+                // counter.fetch_sub(1, AcqRel). Our Acquire load of counter==0
+                // synchronizes-with that fetch_sub, making the abort_now store
+                // visible. Checking in the opposite order (abort_now first) has a
+                // TOCTOU race: an aborting dep can set abort_now and decrement
+                // between our two reads, letting us see (false, 0) and commit.
+                if tx.abort_now.load(Ordering::Acquire) {
+                    return Err(LimboError::CommitDependencyAborted);
+                }
```
Comment (avinassh) [inline core/mvcc/database/mod.rs:1363]: lets do the same like previous one, lets check for `abort_now` first and then proceed to `commit_dep_counter`

File: core/mvcc/database/mod.rs [hunk 6]
```diff
@@ -1068,7 +1112,58 @@ impl<Clock: LogicalClock> StateTransition for CommitStateMachine<Clock> {
...
+
+                // Read-only fast path: if write_set is empty, commit without
+                // going through CommitEnd. CommitEnd updates last_committed_tx_ts
+                // which would make a read-only transaction look like a write,
+                // causing spurious Busy errors from acquire_exclusive_tx.
+                if self.write_set.is_empty() {
```
Comment (avinassh) [inline core/mvcc/database/mod.rs:1369]: let's assert that a read txn always have `commit_dep_set` to zero

File: core/mvcc/database/mod.rs [hunk 7]
```diff
@@ -1116,25 +1211,21 @@ impl<Clock: LogicalClock> StateTransition for CommitStateMachine<Clock> {
                             for row_version in row_versions.iter_mut() {
                                 if let Some(TxTimestampOrID::TxID(id)) = row_version.begin {
                                     if id == self.tx_id {
-                                        // New version is valid STARTING FROM committing transaction's end timestamp
-                                        // See diagram on page 299: https://www.cs.cmu.edu/~15721-f24/papers/Hekaton.pdf
```
Comment (avinassh) [inline core/mvcc/database/mod.rs:1324]: comment removed unnecessarily

File: core/mvcc/database/mod.rs [hunk 8]
```diff
@@ -1116,25 +1211,21 @@ impl<Clock: LogicalClock> StateTransition for CommitStateMachine<Clock> {
...
                                             &mut log_record.row_versions,
                                             row_version.clone(),
-                                        ); // FIXME: optimize cloning out
+                                        );
                                     }
                                 }
                                 if let Some(TxTimestampOrID::TxID(id)) = row_version.end {
                                     if id == self.tx_id {
-                                        // Old version is valid UNTIL committing transaction's end timestamp
-                                        // See diagram on page 299: https://www.cs.cmu.edu/~15721-f24/papers/Hekaton.pdf
```
Comment (avinassh) [inline core/mvcc/database/mod.rs:1336]: comment removed unnecessarily

File: core/mvcc/database/mod.rs [hunk 9]
```diff
@@ -1250,7 +1339,22 @@ impl<Clock: LogicalClock> StateTransition for CommitStateMachine<Clock> {
                 tx_unlocked
                     .state
                     .store(TransactionState::Committed(*end_ts));
+
+                // Hekaton Section 3.3: "The transaction then processes all outgoing
+                // commit dependencies listed in its CommitDepSet. If it committed, it
+                // decrements the target transaction's CommitDepCounter."
```
Comment (avinassh) [inline core/mvcc/database/mod.rs:1568]:
```suggestion
                // Hekaton Section 3.3: "The transaction then processes all outgoing
                // commit dependencies listed in its CommitDepSet. If it committed, it
                // decrements the target transaction's CommitDepCounter."
                // IOW since this txn is successfully committed, let's signal waiting dependent
                // txns about this and reduce their dependency count
```

File: core/mvcc/database/mod.rs [hunk 10]
```diff
@@ -2826,6 +2930,17 @@ impl<Clock: LogicalClock> MvStore<Clock> {
...
         self.unlock_commit_lock_if_held(tx);
 
+        // Hekaton Section 3.3: "If it aborted, it forces the dependent transactions
+        // to also abort by setting their AbortNow flags."
+        let dependents = std::mem::take(&mut *tx.commit_dep_set.lock());
+        for dep_tx_id in dependents {
+            if let Some(dep_tx_entry) = self.txs.get(&dep_tx_id) {
+                let dep_tx = dep_tx_entry.value();
+                dep_tx.abort_now.store(true, Ordering::Release);
+                dep_tx.commit_dep_counter.fetch_sub(1, Ordering::AcqRel);
```
Comment (avinassh) [inline core/mvcc/database/mod.rs:3214]: minor optimisation nit: the transaction is going to abort anyways, no need to update `commit_dep_counter`

File: core/mvcc/database/mod.rs [hunk 11]
```diff
@@ -4244,21 +4363,91 @@ impl RowVersion {
...
+        // CommitDepSet). Dependency is trivially resolved.
+        return;
+    };
+    let depended_on = depended_on.value();
+
+    // Hold lock while checking state to serialize with the drain in
+    // commit/abort postprocessing.
+    let mut dep_set = depended_on.commit_dep_set.lock();
+    match depended_on.state.load() {
+        TransactionState::Preparing(_) | TransactionState::Active => {
```
Comment (avinassh) [inline core/mvcc/database/mod.rs]:
```suggestion
        TransactionState::Active => {
             turso_assert!(false, "a txn found dependent on a active txn");
        },
        TransactionState::Preparing(_) =>
```

⚠️ ⚠️  This match arm is incorrect, I suggest keeping an assert

File: core/mvcc/database/mod.rs [hunk 12]
```diff
@@ -4244,21 +4363,91 @@ impl RowVersion {
...
+            let tb = tb_entry.value();
             let visible = match tb.state.load() {
                 TransactionState::Active => tx.tx_id == tb.tx_id && rv.end.is_none(),
-                TransactionState::Preparing(_) => false, // NOTICE: makes sense for snapshot isolation, not so much for serializable!
+                TransactionState::Preparing(end_ts) => {
+                    // Hekaton Table 1 / Section 2.5: speculative read of TB.
+                    // If begin_ts >= end_ts, the version would be visible once TB
+                    // commits. Speculatively return true and register a dependency.
+                    // Fixes partial commit visibility (Bug #8).
+                    if tx.begin_ts >= end_ts && tx.tx_id != tb.tx_id {
```
Comment (avinassh) [inline core/mvcc/database/mod.rs:4828]:
```suggestion
                    turso_assert!(tx.tx_id != tb.tx_id, "a txn cannot read its own row versions during prepare");
                    if tx.begin_ts >= end_ts {
```
I am sus about second check. `tx.tx_id != tb.tx_id` is this needed? Rather I'd assert that they are not same. because a txn in preparing phase won't read row versions right

File: core/mvcc/database/mod.rs [hunk 13]
```diff
@@ -4281,17 +4470,27 @@ fn is_end_visible(
...
                 // Source: https://avi.im/blag/2023/hekaton-paper-typo/
                 TransactionState::Active => current_tx.tx_id != other_tx.tx_id,
-                // Table 2 (Hekaton): If TS > RT, V is visible. If TS < RT, T speculatively ignores V.
-                TransactionState::Preparing(end_ts) => current_tx.begin_ts < end_ts,
+                // Hekaton Table 2: speculative ignore of TE. If end_ts <= begin_ts,
+                // we speculatively ignore V (treat deletion as committed). Register a
+                // dependency in case TE aborts (then V should have been visible).
+                TransactionState::Preparing(end_ts) => {
+                    let visible = current_tx.begin_ts < end_ts;
+                    if !visible && current_tx.tx_id != other_tx.tx_id {
```
Comment (avinassh) [inline core/mvcc/database/mod.rs]:
```suggestion
                    turso_assert!(current_tx.tx_id != other_tx.tx_id, "a txn cannot read its own row versions during prepare");
                    if end_ts <= current_tx.begin_ts {
```

same like earlier and also simplified the if

File: core/mvcc/database/tests.rs [hunk 14]
```diff
@@ -5806,3 +6637,200 @@ fn test_close_persists_drop_index() {
...
+         \n\
+         ROOT CAUSE: Commit loop (mod.rs:912-984) updates row timestamps non-atomically\n\
+         while transaction state remains Preparing. Concurrent readers see inconsistent\n\
+         snapshots with partial transaction visibility.\n\
+         \n\
+         FIX REQUIRED: Make timestamp updates atomic, or change visibility logic to\n\
+         always dereference transaction state instead of reading row timestamps directly."
+        );
+    }
+}
```
Comment (avinassh) [inline core/mvcc/database/tests.rs:6836]: tests are nice 👌 👌 👌

File: core/mvcc/database/mod.rs
Comment (avinassh) [inline core/mvcc/database/mod.rs:3068]:
```suggestion
pub fn remove_tx(&self, tx_id: TxID) {
    if let Some(entry) = self.txs.get(&tx_id) {
        let tx = entry.value();
        let dep_set = tx.commit_dep_set.lock();
        // Invariant: commit_dep_set must be drained before removal.
        // CommitEnd and rollback_tx both drain this set to notify dependents
        // (decrement their CommitDepCounter / set AbortNow). If we remove a tx
        // with a non-empty CommitDepSet, those dependents will wait forever in
        // WaitForDependencies — a silent deadlock.
        turso_assert!(
            dep_set.is_empty(),
            "remove_tx({tx_id}): commit_dep_set is not empty ({} dependents would be orphaned)",
            dep_set.len(),
        );
    }

    self.txs.remove(&tx_id);
    self.blocking_checkpoint_lock.unlock();
}
```

WDYT?

**Comment by avinassh** [comment]
Todos for me, because no need to wait for this PR to merged:

- [ ] verify / validate that a txn never remains in a `WaitForDependencies` state
- [ ] verify / validate that we can never have cyclic dependencies 
- [ ] verify / validate that we can _never_ be dependent on a future transaction

File: core/mvcc/database/mod.rs (hunk 12 above)
Comment (avinassh) [reply core/mvcc/database/mod.rs:4828]: @pereman2 you likely missed this one

File: core/mvcc/database/mod.rs (hunk 6 above)
Comment (avinassh) [reply core/mvcc/database/mod.rs:1369]: @pereman2 this too
